apis/r/vignettes/soma-experiment-queries.Rmd

```diff
@@ -66,17 +66,85 @@ Number of variables:
 ```
 
-We can also extract any data component from the experiment. Here we'll read in the `obs` data frame from the query.
+We can also extract any data component from the experiment. Here we'll read in the `obs` 
+data frame from the query using `obs()` which returns an iterator of `arrow::Table`. The iterator is useful 
+when the data is too large to load in memory allowing to stream the data in chunks. This 
+applies to `var()` as well.
+
+To load the data in memory you can concatenate all chunks of the iterator as shown below.
 
 ```{r}
-query$obs()
+iterator <- query$obs()
+obs <- iterator$concat()
+obs
+```
+As a reminder `arrow:Table` can be easily cast into a `tibble`
+
+```{r}
+as.data.frame(obs)
+```
+
+Alternatively, you can use the iterator, which retrieves data in chunks that are smaller than 
+the `soma.init_buffer_bytes` context field. You can use the iterator's method `$read_next()` to 
+load a chunk in memory. In this example the data is so small that it all fits in one chunk.
+
+```{r}
+iterator <- query$obs()
+iterator$read_next()
+```
+
+In this example the full `obs` table is relatively small and fits all in one chunk.
+
+For a bigger `SOMADataFrame` you can check if the iteration has finished by checking the logical `$read_complete()`.
+
+Here we demonstrate by creating a new iterator.
+
+```{r}
+iterator <- experiment$obs$read()
+iterator$read_complete()
+```
+```{r}
+iterator$read_next()
+iterator$read_complete()
+```
+
+We can also access the expression via `X()`. 
+
+Similarly to `obs()` and `var()`, `X()` is intended for iteration, but in this case we have 
+access to two different iterators, and thus `X()` returns a reader that gives you access to an 
+iterator for `arrow::Table` and one for `Matrix::sparse_matrix`.
+
+Let's take a look at the Arrow Table iterator:
+
+```{r}
+reader <- query$X(layer_name = "counts")
+table_irerator <- reader$tables()
+table_irerator$read_next()
+```
+
+As in the `obs` example the data is small enough to fit in one chunk. For bigger data
+you can user `iterator$read_complete()` to check the status of iteration and `iterator$concat()`
+to concatenate the rest of the chunks.
+
+The iterator for  `Matrix::sparse_matrix` works in the same way. Keep in mind that the
+matrix format is `dgTMatrix` as it is the most memory-efficient and the only
+format type that can be easily iterated. And most importantly, the resulting
+object is a "view" of the full matrix with the original shape and indexes but only
+with data corresponding to the query coordinates or filters (see section below).
+
+```{r}
+reader <- query$X(layer_name = "counts")
+iterator <- reader$sparse_matrix()
+str(iterator$read_next())
 ```
 
 ## Adding filters
 
-Adding filters requires creating a `SOMAAxisQuery` object that allows you to define coordinates, value filters, or both for an axis.
+Adding filters requires creating a `SOMAAxisQuery` object that allows you to define coordinates, 
+value filters, or both for an axis.
```

Comment:
Looks like your editor is hard-wrapping markdown text.

Reply:
I'll check my config settings. I use Rstudio for the vignettes, nvim for everything else.